src/server/ocpp/services/OCPPService.ts

```diff
@@ -1135,31 +1135,36 @@ export default class OCPPService {
 
   private async checkNotificationEndOfCharge(tenant: Tenant, chargingStation: ChargingStation, transaction: Transaction) {
     // Transaction in progress?
-    if (transaction && !transaction.stop) {
-      // Has consumption?
-      if (transaction.numberOfMeterValues > 1 && transaction.currentTotalConsumptionWh > 0) {
-        // End of charge?
-        if (this.chargingStationConfig.notifEndOfChargeEnabled && transaction.currentTotalConsumptionWh > 0) {
-          // Battery full
-          if (transaction.currentStateOfCharge === 100) {
+    if (!transaction?.stop && transaction.currentTotalConsumptionWh > 0) {
+      // Check the battery
+      if (transaction.currentStateOfCharge > 0) {
+        // Check if battery is full (100%)
+        if (transaction.currentStateOfCharge === 100) {
+          // Send Notification
+          this.notifyEndOfCharge(tenant, chargingStation, transaction);
+        // Check if optimal charge has been reached (85%)
+        } else if (transaction.currentStateOfCharge >= this.chargingStationConfig.notifBeforeEndOfChargePercent) {
+          // Send Notification
+          this.notifyOptimalChargeReached(tenant, chargingStation, transaction);
+        }
+      // No battery information: check last consumptions
+      } else {
+        // Connector status must be 'Charging'
+        const connector = Utils.getConnectorFromID(chargingStation, transaction.connectorId);
+        if (connector.status === ChargePointStatus.CHARGING) {
+          // Check the last 3 consumptions
+          const consumptions = await ConsumptionStorage.getTransactionConsumptions(
+            tenant.id, { transactionId: transaction.id }, { limit: 3, skip: 0, sort: { startedAt: -1 } });
+          if (consumptions.count === 3) {
+            // Check the consumptions
+            const noConsumption = consumptions.result.every((consumption) =>
+              consumption.consumptionWh === 0 &&
+              consumption.limitSource !== ConnectorCurrentLimitSource.CHARGING_PROFILE);
```

Comment:
I think we are missing a check here. Normally we have a charging profile in place when the car is fully charged. 
So let us say we have a charging profile in place with a limit of 11kW, but the car is not charging anymore. 
The consumptionWh, will be 0, but the limitSource will be the charging profile so no notification is sent, even if the car finished charging.